go/store/types/value_store.go

```diff
@@ -253,34 +252,23 @@ func (lvs *ValueStore) ReadManyValues(ctx context.Context, hashes hash.HashSlice
 	}
 
 	if len(remaining) != 0 {
-		// Request remaining hashes from ChunkStore, processing the found chunks as they come in.
-		foundChunks := make(chan *chunks.Chunk, 16)
-
-		ae := atomicerr.New()
-		go func() {
-			defer close(foundChunks)
-			err := lvs.cs.GetMany(ctx, remaining, foundChunks)
-			ae.SetIfError(err)
-		}()
-
-		var err error
-		for c := range foundChunks {
-			if err != nil {
-				continue // continue to drain even if there is an error
+		mu := new(sync.Mutex)
+		var decodeErr error
+		err := lvs.cs.GetMany(ctx, remaining, func(c *chunks.Chunk) {
```

Comment:
A potential downside of this callback change is that callers have to be more aware of thread safety. Here I introduced a race on the writes to `decodeErr` and `foundValues` before realizing it and adding the mutex.